### PR TITLE
Laser beam

### PR DESCRIPTION
**Pull request by hansieodendaal**
This is a new report. All detail equations have been constructed based on published high-level information from Beam and discussions with the Beam team.

Laser Beam is Beam's version of the Lightning Network for Mimblewimble.

### CODE REVIEW

File: src/scaling/laser-beam/MainReport.md [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+
+   In the event that for round $N$, Alice or Bob decides to stop negotiations after their respective part 1 and part 2 
+   have been concluded and those transactions have been broadcast, it will effectively be a channel closure.
+
+1. Revoke Attack Vector
+
+   When revoking the previous refund  $(N-1)$, Alice can get hold of Bob's blinding factor share $\hat{k}\_{(N-1)\_{b}}$ 
+   (17), and after verifying that it is correct (18), refuse to give up her blinding factor share. This will leave 
+   Alice with the ability to broadcast any of refund transactions $(N-1)$ and $N$, without fear of Bob broadcasting a 
+   punishment transaction. Bob, on the other hand, will only be able to broadcast refund transaction $N$.
```
Comment (SWvheerden) [inline src/scaling/laser-beam/MainReport.md]: This is not a problem. Lighting has the same issue but solves it with the way the messages are sent:
https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md#normal-operation

File: src/scaling/laser-beam/MainReport.md [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+Lightning Network's dispute mechanism requires all users to constantly watch the blockchain for fraud. Various mainnet 
+implementations that support Bitcoin exist and, with small tweaks, some of them are also able to support Litecoin 
+([[2]], [[3]], [[4]], [[10]]).
+
+Laser Beam is an adaptation of the Lightning Network for the [Mimblewimble](../../protocols/mimblewimble-1/MainReport.md) 
+protocol, to be implemented for Beam ([[5]], [[6]], [[7]]). At the time of writing of this report (November&nbsp;2019), 
+the specifications were far advanced, but still work in progress. Beam has a working demonstration in its mainnet 
+repository, which at this stage demonstrates off-chain transactions in a single channel between two parties [[8]]. 
+According to the Request for Comment (RFC) documents, Beam does not plan to support multiparty (more than two) payment 
+channels, but rather to implement routing across different payment channels in the Lightning Network style.
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Good catch! Agreed

File: src/scaling/laser-beam/MainReport.md [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+each user creates two transactions: one kept partially secret (discussed below) and the other shared. Each partially 
+secret transaction creates a different intermediate multiparty UTXO, which is then used as input in two shared 
+transactions, to spending the same set of outputs to each participant.
+
+All consecutive refund procedures work in exactly the same way. In the equations that follow, double subscripts 
+$\_{AA}$, $\_{AB}$, $\_{BA}$ and $\_{BB}$ have the following meaning: the first letter and the second letter indicate 
+who controls the transaction and who created the value, respectively. Blinding factors denoted by 
+$\hat{k}$ and $k^{\prime}$, and values denoted by $v^{\prime}$, have a special purpose, discussed later, so that 
+$\hat{k}\_{N\_{a}} \neq k\_{N\_{a}} \neq k^{\prime}\_{N\_{a}}$ and $v\_{N\_{a}} \neq v^{\prime}\_{N\_{a}}$. The 
+$N\_{\text{th}}$ refund procedure is as follows:
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Graph is a very good idea

File: src/scaling/laser-beam/MainReport.md [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+$$
+\begin{aligned}
+  \text{Signature tuple: }(s\_{N\_{AA1}}+s\_{N\_{AB1}},R\_{N\_{AA1}}+R\_{N\_{AB1}}) 
+\end{aligned}
+\mspace{50mu} (5)
+$$
+
+$$
+\begin{aligned}
+  \text{Kernel of this transaction, kept secret:}\quad\mathcal{K}\_{N\_{AA1}}
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Alice collects Bob's half of the signature adds her half of the signature for transaction part 1, thus completes the kernel, but she does not share it.

File: src/scaling/laser-beam/MainReport.md [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+$$
+\begin{aligned}
+  \Big(v\_{0}H+(\hat{k}\_{(N-1)\_{a}}+k\_{(N-1)\_{b}})G\Big) \overset{?}{=} C(v\_{0},\ \hat{k}\_{(N-1)\_{a}}+k\_{(N-1)\_{b}})
+\end{aligned}
+\mspace{50mu} (19)
+$$
+
+
+Although each party will now have its counterparty's blinding factor share in the counterparty's intermediate multiparty 
+UTXO, they will still not be able to spend it, because the corresponding transaction kernel is still kept secret by the
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Clarified

File: src/scaling/laser-beam/MainReport.md [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+UTXO, they will still not be able to spend it, because the corresponding transaction kernel is still kept secret by the 
+counterparty.
+
+### Punishment Transaction
+
+If a counterparty decides to broadcast a revoked set of refund transactions, and if the honest party is actively monitoring 
+the blockchain and able to detect the attempted foul play, a punishment transaction can immediately be constructed 
+before the relative time lock $h_{rel}$ expires. Whenever any of the counterparty's intermediate multiparty UTXOs, 
+$\text{MultiSig}(N-m)\ \text{for}\ 0<m<N$, becomes available in the blockchain, the honest party can spend all the 
+funds to its own output, because it knows the total blinding factor.
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Clarified

File: src/scaling/laser-beam/MainReport.md [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+
+### Channel Closure
+
+Whenever the parties agree to a channel closure, the original on‑chain multiparty UTXO, $\text{MultiSig}(0)​$, is spent 
+to their respective outputs in a collaborative transaction. In the event of a single party deciding to close the channel 
+unilaterally for whatever reason, its latest refund transaction is broadcast, effectively closing the channel.
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Clarified

File: src/scaling/laser-beam/MainReport.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (SWvheerden) [reply src/scaling/laser-beam/MainReport.md]: Yes thats true, but the whole kernel is not kept secret. Part of it is missing for Bob (Alice's signature).

File: src/scaling/laser-beam/MainReport.md [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+
+## Conclusions, Observations and Recommendations
+
+1. MultiSig
+
+   The Laser Beam MultiSig corresponds to a Mimblewimble $2-of-2$ Multiparty Bulletproof UTXO, as described 
+   [here](../../protocols/mimblewimble-mp-bp-utxo/MainReport.md#mimblewimble--ntext-of-n--multiparty-bulletproof-utxo). 
+   There is more than one method for creating the MultiSig's associated Bulletproof range proof, and utilizing the 
+   [Bulletproofs MPC Protocol](../../protocols/mimblewimble-mp-bp-utxo/MainReport.md#creating-the-multiparty-bulletproof-range-proof) 
+   can make it more secure.
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: It is an observation. Changed `can make it more secure` to `may make it more secure and enable wallet reconstruction`.

File: src/scaling/laser-beam/MainReport.md [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+
+1. MultiSig
+
+   The Laser Beam MultiSig corresponds to a Mimblewimble $2-of-2$ Multiparty Bulletproof UTXO, as described 
+   [here](../../protocols/mimblewimble-mp-bp-utxo/MainReport.md#mimblewimble--ntext-of-n--multiparty-bulletproof-utxo). 
+   There is more than one method for creating the MultiSig's associated Bulletproof range proof, and utilizing the 
+   [Bulletproofs MPC Protocol](../../protocols/mimblewimble-mp-bp-utxo/MainReport.md#creating-the-multiparty-bulletproof-range-proof) 
+   can make it more secure. 
+
+1. Linked Transactions
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: In Markdown a list can just be `1. ,,,,` `1. ,,,,`  `1. ,,,,`  `1. ,,,,`. It renders correctly.

File: src/scaling/laser-beam/MainReport.md [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+
+   Part 2 of the refund procedure requires a kernel with a relative time lock to the kernel of its corresponding 
+   part 1 refund procedure, when those kernels are not yet available in the base layer, as well as a different, 
+   non-standard, signature challenge. Metadata about the linked transaction kernel and non-standard signature challenge
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: The essence is the challenge for Part 1 and Part 2 must be calculated differently. If Part is the standard method, Part 2 is not.

Changed `non-standard signature challenge` to `non-similar signature challenge`

File: src/scaling/laser-beam/MainReport.md [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+   can make it more secure. 
+
+1. Linked Transactions
+
+   Part 2 of the refund procedure requires a kernel with a relative time lock to the kernel of its corresponding 
+   part 1 refund procedure, when those kernels are not yet available in the base layer, as well as a different, 
+   non-standard, signature challenge. Metadata about the linked transaction kernel and non-standard signature challenge 
+   creation must therefore be embedded within part 2 refund transaction kernels.  
+
+1. Refund Procedure
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: See previous comment

File: src/scaling/laser-beam/MainReport.md [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+   In the event that for round $N$, Alice or Bob decides to stop negotiations after their respective part 1 has been 
+   concluded and that transaction has been broadcast, the result would be that funding UTXO, $\text{MultiSig}(0)$, would be 
+   replaced by the respective $\text{MultiSig}(N)$. The channel will still be open. However, it also cannot be spent 
+   unilaterally, as the blinding factor is shared. Any new updates of the channel will then need to be based on 
+   $\text{MultiSig}(N)$ as the funding UTXO.
+
+   In the event that for round $N$, Alice or Bob decides to stop negotiations after their respective part 1 and part 2 
+   have been concluded and those transactions have been broadcast, it will effectively be a channel closure.
+
+1. Revoke Attack Vector
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: See previous comment in this regard

File: src/scaling/laser-beam/MainReport.md (hunk 1 above)
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Confirmed with Beam (Vladislav G) that this is a possibility; one party can decide not to reveal its blinding factor share after the counterparty's share has been received.

File: src/scaling/laser-beam/MainReport.md [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+   Part 2 of the refund procedure requires a kernel with a relative time lock to the kernel of its corresponding 
+   part 1 refund procedure, when those kernels are not yet available in the base layer, as well as a different, 
+   non-standard, signature challenge. Metadata about the linked transaction kernel and non-standard signature challenge 
+   creation must therefore be embedded within part 2 refund transaction kernels.  
+
+1. Refund Procedure
+
+   In the event that for round $N$, Alice or Bob decides to stop negotiations after their respective part 1 has been 
+   concluded and that transaction has been broadcast, the result would be that funding UTXO, $\text{MultiSig}(0)$, would be 
+   replaced by the respective $\text{MultiSig}(N)$. The channel will still be open. However, it also cannot be spent
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Clarified, added that logic.

File: src/scaling/laser-beam/MainReport.md [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+- [Conclusions, Observations and Recommendations](#conclusions-observations-and-recommendations)
+- [Appendices](#appendices)
+  - [Appendix A: Notation Used](#appendix-a-notation-used)
+- [References](#references)
+- [Contributors](#contributors)
+
+
+## Introduction
+
+Proof-of-Work (PoW) blockchains are slow and have poor scalability properties. A payment channel
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Updated

File: src/scaling/laser-beam/MainReport.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Ok, got you, final signature is kept secret by not sharing the final kernel.

File: src/scaling/laser-beam/MainReport.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (SWvheerden) [reply src/scaling/laser-beam/MainReport.md]: I still disagree on the secure part. Why do you say its more secure? I can see the wallet reconstruction as being something. But that does not make it more secure?

File: src/scaling/laser-beam/MainReport.md (hunk 1 above)
Comment (SWvheerden) [reply src/scaling/laser-beam/MainReport.md]: Mmm, Well they should be able to solve it by just doing it the same way lightning does it as in lightning also needs to share the pre-image hashes with each other.

File: src/scaling/laser-beam/MainReport.md [hunk 15]
```diff
@@ -0,0 +1,475 @@
...
+
+Alice and Bob set up Alice's intermediate MultiSig funding transaction (Figure&nbsp;1), spending the original funding 
+MultiSig UTXO. The lock height $h\_{N}$ corresponds to the current blockchain height. 
+
+$$
+\begin{aligned}
+  -\text{MultiSig}(0)+\text{MultiSig}(N)\_{A}+\text{fee} 
+       & =\text{Excess}(N)\_{A1} \\\\
+  -\Big(v\_{0}H+(k_{0\_{a}}+k\_{0\_{b}})G\Big) + \Big(v\_{0}H+(\hat{k}\_{N\_{a}}+k\_{N\_{b}})G\Big) + fH 
+       &= \mathcal{X}\_{N\_{A1}}
```
Comment (Empiech007) [inline src/scaling/laser-beam/MainReport.md]: What assumptions are imposed on the blinded fee  "fH"  and why does it not appear in equation (3) ?

File: src/scaling/laser-beam/MainReport.md [hunk 16]
```diff
@@ -0,0 +1,475 @@
...
+  \text{Final signature tuple, kept secret:}\quad(s\_{N\_{AA1}}+s\_{N\_{AB1}},R\_{N\_{AA1}}+R\_{N\_{AB1}}) 
+\end{aligned}
+\mspace{50mu} (5)
+$$
+
+$$
+\begin{aligned}
+  \text{Kernel of this transaction:}\quad\mathcal{K}\_{N\_{AA1}} 
+\end{aligned}
+\mspace{50mu} (6)
```
Comment (Empiech007) [inline src/scaling/laser-beam/MainReport.md]: It is not clear why this is an equation on its own. It appears to be simply 'stating notation', unless it spells out what the actual kernel would be in this case. ... same comment for equation (14).

File: src/scaling/laser-beam/MainReport.md (hunk 15 above)
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Good catch! In equation 2 and 10 the values on H should be `-v_0 + (v_0-f) + f = 0`
Note that `fH` is not considered blinded as it is equal to Pedersen commitment `fH + 0G`, zero blinding factor.

File: src/scaling/laser-beam/MainReport.md (hunk 16 above)
Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: Yes, those are not equations on its own, but important to define as those values are used in (9) and (17) respectively.

File: src/scaling/laser-beam/MainReport.md [hunk 17]
```diff
@@ -0,0 +1,475 @@
...
+  -\text{MultiSig}(0)+\text{MultiSig}(N)\_{A}+\text{fee} 
+       & =\text{Excess}(N)\_{A1} \\\\
+  -\Big(v\_{0}H+(k_{0\_{a}}+k\_{0\_{b}})G\Big) + \Big((v\_{0}-f)H+(\hat{k}\_{N\_{a}}+k\_{N\_{b}})G\Big) + fH 
+       &= \mathcal{X}\_{N\_{A1}} 
+\end{aligned}
+\mspace{50mu} (2)
+$$
+
+They collaborate to create $\text{MultiSig}(N)\_{A}$, its Bulletproof range proof,
+the signature challenge and Bob's potion of the signature, $s\_{N\_{AB1}}$. Alice does not share the final kernel and
```
Comment (mikethetike) [inline src/scaling/laser-beam/MainReport.md]:
```suggestion
the signature challenge and Bob's portion of the signature, $s\_{N\_{AB1}}$. Alice does not share the final kernel and 
```

File: src/scaling/laser-beam/MainReport.md [hunk 18]
```diff
@@ -0,0 +1,475 @@
...
+- [Conclusions, Observations and Recommendations](#conclusions-observations-and-recommendations)
+- [Appendices](#appendices)
+  - [Appendix A: Notation Used](#appendix-a-notation-used)
+- [References](#references)
+- [Contributors](#contributors)
+
+
+## Introduction
+
+Proof-of-Work (PoW) blockchains are not immutable in realtime and have poor scalability properties. A payment channel
```
Comment (mikethetike) [inline src/scaling/laser-beam/MainReport.md]:
```suggestion
Proof-of-Work (PoW) blockchains have long confirmation times and are not suited to realtime transactions. A payment channel 
```

Comment (SWvheerden) [reply src/scaling/laser-beam/MainReport.md]: Long confirmations is relative as well. 
Perhaps we should rewrite this as:
```suggestion
Proof-of-Work (PoW) blockchains are not suited to realtime transactions as there is no guarantee that the transaction will end up on the blockchain . A payment channel 
```

Comment (hansieodendaal) [reply src/scaling/laser-beam/MainReport.md]: The original meaning is lost, will change as follows:
`Proof-of-Work (PoW) blockchains are notoriously slow, as transactions need to be a number of blocks in the past to be confirmed, and have poor scalability properties.`

**Comment by hansieodendaal** [comment]
Thank you @SWvheerden, @Empiech007 and @mikethetike for your review comments!

**Pull request merged**